course02/project01/gavin/Asteroids/game_store.h

```diff
@@ -0,0 +1,61 @@
+#include "utils.h"
+#include <tuple>
+#include <vector>
+#include <algorithm>
+#include <unordered_map>
+/*
+ * Holds a vector of each provided type, along with a lookup table to provide stable access to the vector elements.
+ */
+template<class ...Objs>
+class GameStore {
+public:
+
+    //Reallocating the vector would invalidate the lookup table.
+    GameStore() {
+        (_Reserve<Objs>(1024), ...);
+    }
+
+
+    template<class Func>
+    void ForAll(Func f);
+
+    template<class Type, class Func>
+    void ForAllOf(Func f);
+
+
+    template<class Func>
+    void RemoveIf(Func f);
+
+    template<class Obj, class ...Args>
+    ObjectId Create(Args ...args) {
+        Obj& newObject = std::get<std::vector<Obj>>(mContents).emplace_back(args...);
+        std::get<std::unordered_map<ObjectId, Obj*>>(mLookup).insert({newObject.GetId(), &newObject});
+        return newObject.GetId();
+    }
+
+    template<class Object>
+    Object& Get(ObjectId id)
+    {
+        return *std::get<std::unordered_map<ObjectId, Object*>>(mLookup)[id];
+    }
+
+private:
+
+    std::tuple<std::vector<Objs>...> mContents;
+    std::tuple<std::unordered_map<ObjectId, Objs*>...> mLookup;
```

Comment:
Not a big fan of using a unique id for my lookup, alternatively I suspect it may be possible for elements to update their own table entries when they are moved/deleted. However at this point my IDE is no longer even linting this code correctly.